lib/neo4j/shared/declared_property_manager.rb

```diff
@@ -0,0 +1,94 @@
+module Neo4j::Shared
+  class DeclaredPropertyManager
+    include Neo4j::Shared::TypeConverters
+
+    attr_reader :klass
+
+    def initialize(klass)
+      @klass = klass
+    end
+
+    def register(property)
+      @_attributes_nil_hash = nil
+      registered_properties[property.name] = property
+      register_magic_typecaster(property) if property.magic_typecaster
+      declared_property_defaults[property.name] = property.default_value if property.default_value
+    end
+
+    def declared_property_defaults
+      @_default_property_values ||= {}
+    end
+
+    def registered_properties
+      @_registered_properties ||= {}
+    end
+
+    def attributes_nil_hash
+      @_attributes_nil_hash ||= {}.tap { |attr_hash| registered_properties.each_pair { |k, _v| attr_hash[k.to_s] = nil } }.freeze
```

Comment:
:each_key creates a new array, deliberately avoiding it wherever possible.

On Friday, May 1, 2015, Brian Underwood <notifications@github.com> wrote:

> In lib/neo4j/shared/declared_property_manager.rb
> <https://github.com/neo4jrb/neo4j/pull/773#discussion_r29535767>:
>
> > +      @_attributes_nil_hash = nil
> > +      registered_properties[property.name] = property
> > +      register_magic_typecaster(property) if property.magic_typecaster
> > +      declared_property_defaults[property.name] = property.default_value if property.default_value
> > +    end
> > +
> > +    def declared_property_defaults
> > +      @_default_property_values ||= {}
> > +    end
> > +
> > +    def registered_properties
> > +      @_registered_properties ||= {}
> > +    end
> > +
> > +    def attributes_nil_hash
> > +      @_attributes_nil_hash ||= {}.tap { |attr_hash| registered_properties.each_pair { |k, _v| attr_hash[k.to_s] = nil } }.freeze
>
> Perhaps _ instead of _v to show that it's really not used?
>
> Also, you could use Hash#each_key
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/neo4jrb/neo4j/pull/773/files#r29535767>.
>